crates/core/tedge_mapper/src/core/mapper.rs

```diff
@@ -140,13 +147,62 @@ impl Mapper {
             let _ = self.output.send(health_message).await;
         } else {
             let converted_messages = self.converter.convert(&message).await;
+
             for converted_message in converted_messages.into_iter() {
                 let _ = self.output.send(converted_message).await;
             }
         }
     }
 }
 
+async fn process_inotify_and_mqtt_messages(
+    mapper: &mut Mapper,
+    dir: PathBuf,
+) -> Result<(), MqttError> {
+    match create_inofity_event_stream(dir.clone()) {
+        Ok(mut inotify_events) => loop {
+            tokio::select! {
+                msg =  mapper.input.next() => {
+                    match msg {
+                        Some(message) => {
+                            mapper.process_message(message).await;
+                        } None => {
+                            break Ok(());
+                        }
+                    }
+                }
+                event = inotify_events.next() => {
+                    match event {
+                        Some(ev) => {
+                            match ev {
+                                Ok(ev_string) => {
+
+                                            match  process_inotify_events(dir.clone(), ev_string) {
+                                                Ok(discovered_ops) => {
+                                                    let _ = mapper.output.send(mapper.converter.process_operation_update_message(discovered_ops)).await;
+                                                }
+                                                Err(e) => {eprintln!("Processing inotify event failed due to {}", e);}
+                                            }
+
+
+                                } Err(e) => {eprintln!("Failed to extract event {}", e);}
+                            }
+                        }
+                        None => {}
+                    }
+                }
+            }
+        }, // On error continue to process only mqtt messages.
+        Err(e) => {
+            eprintln!("Failed to create the inotify stream due to {:?}. So, dynamic operation discovery not supported, please restart the mapper on Add/Removal of an operation", e);
+            while let Some(message) = mapper.input.next().await {
+                mapper.process_message(message).await;
+            }
+            Ok(())
+        }
```

Comment:
Yes this sounds to be the best approach.

However, since the message processing loop being in two place, it's better to introduce a `process_mqtt_message` function.